src/sqlfluff/parser/grammar.py

```diff
@@ -496,9 +502,13 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context,
                         code_only=code_only)
 
                     if match:
-                        if matcher in start_brackets:
+                        # NB: We can only consider this as a nested bracket if the start
+                        # and end tokens are not the same. If a matcher is both a start and
+                        # end token we cannot deepen the bracket stack.
+                        if matcher in start_brackets and matcher not in end_brackets:
                             # Same procedure as below in finding brackets.
                             bracket_stack.append(match.matched_segments[0])
+                            definite_stack.append(start_definite[start_brackets.index(matcher)])
```

Comment:
It looks like `bracket_stack` and `definite_stack` are always updated in unison. It might be easier to understand if you combined them into a single list of dictionaries, objects, whatever.